### PR TITLE
Fix BUILD_CAFFE2 if FBGEMM and NNPACK are not built

### PR DESCRIPTION
**Pull request by suo**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#45610 Fix BUILD_CAFFE2 if FBGEMM and NNPACK are not built**

Also add to the usual documentation places that this option exists.

Differential Revision: [D24058199](https://our.internmc.facebook.com/intern/diff/D24058199)

### CODE REVIEW

File: caffe2/utils/CMakeLists.txt [hunk 1]
```diff
@@ -1,9 +1,13 @@
 if((NOT BUILD_CAFFE2) OR (INTERN_BUILD_MOBILE AND NOT BUILD_CAFFE2_MOBILE))
   list(APPEND Caffe2_CPU_SRCS
     utils/string_utils.cc
-    utils/threadpool/pthreadpool-cpp.cc
     utils/threadpool/ThreadPool.cc
   )
+
+  if(USE_PTHREADPOOL)
+    list(APPEND Caffe2_CPU_SRCS utils/threadpool/pthreadpool-cpp.cc)
+  endif()
```
Comment (walterddr) [inline caffe2/utils/CMakeLists.txt:9]: we should probably do:
```suggestion
  if(USE_PTHREADPOOL AND NOT USE_INTERNAL_PTHREADPOOL_IMPL)
    list(APPEND Caffe2_CPU_SRCS utils/threadpool/pthreadpool-cpp.cc)
  endif()
```

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -489,8 +489,7 @@ only interested in a specific component.
 - Working on a test binary? Run `(cd build && ninja bin/test_binary_name)` to
   rebuild only that test binary (without rerunning cmake). (Replace `ninja` with
   `make` if you don't have ninja installed).
-- Don't need Caffe2?  Pass `BUILD_CAFFE2_OPS=0` to disable build of
-  Caffe2 operators.
+- Don't need Caffe2?  Pass `BUILD_CAFFE2=0` to disable Caffe2 build.
```
Comment (walterddr) [inline CONTRIBUTING.md:492]: BUILD_CAFFE2=0 also disable other stuff. should we keep both in the doc? (IMO we can put doc optimization in a separate PR?)

**Review by malfet — APPROVED** [review]

**Review by walterddr — APPROVED** [review]

File: CONTRIBUTING.md (hunk 2 above)
Comment (suo) [reply CONTRIBUTING.md:492]: So BUILD_CAFFE2 is not a superset of BUILD_CAFFE2_OPs? If so, it should be noted in this contributing doc.

Comment (walterddr) [reply CONTRIBUTING.md:492]: it is actually. what I meant is to keep both lines

Comment (walterddr) [reply CONTRIBUTING.md:492]: e.g. 
```suggestion
- Don't need Caffe2 at all?  Pass `BUILD_CAFFE2=0` to disable Caffe2 build.
- Don't need Caffe2 ops?  Pass `BUILD_CAFFE2_OPS=0` to disable Caffe2 operator build.
```

Comment (walterddr) [reply CONTRIBUTING.md:492]: although I am a bit hesitate to put this on the CONTRIBUTING.md - BUILD_CAFFE2=OFF doesn't necessarily disable ALL caffe2 builds at the moment, there are still some separation work to be done. 

I wouldn't block merging the PR though.
